pkg/apis/sources/v1beta1/ping_conversion.go

```diff
@@ -18,17 +18,130 @@ package v1beta1
+	return json.Marshal(objmap)
+}
+
 // ConvertTo implements apis.Convertible
-func (source *PingSource) ConvertTo(ctx context.Context, sink apis.Convertible) error {
-	return fmt.Errorf("v1beta1 is the highest known version, got: %T", sink)
+// Converts source from v1beta1.PingSource into a higher version.
+func (source *PingSource) ConvertTo(ctx context.Context, obj apis.Convertible) error {
+	switch sink := obj.(type) {
+	case *v1beta2.PingSource:
+		sink.ObjectMeta = source.ObjectMeta
+		sink.Status = v1beta2.PingSourceStatus{
+			SourceStatus: source.Status.SourceStatus,
+		}
+
+		// deep copy annotations to avoid mutation on source.ObjectMeta.Annotations
+		annotations := make(map[string]string)
+		for key, value := range source.GetAnnotations() {
+			annotations[key] = value
+		}
+
+		// try to unmarshal v1beta2.PingSource.Spec from V1B2SpecAnnotationKey
+		v1beta2Spec, ok := annotations[V1B2SpecAnnotationKey]
+		if ok {
+			if err := json.Unmarshal([]byte(v1beta2Spec), &sink.Spec); err != nil {
+				ok = false
+			}
+			// we don't need this annotation in a v1beta2.PingSource object
+			delete(annotations, V1B2SpecAnnotationKey)
+		}
+
+		// cannot unmarshal, do a normal conversion
+		if !ok {
+			sink.Spec = v1beta2.PingSourceSpec{
+				SourceSpec: source.Spec.SourceSpec,
+				Schedule:   source.Spec.Schedule,
+				Timezone:   source.Spec.Timezone,
+			}
+
+			if source.Spec.JsonData != "" {
+				msg, err := makeMessage(source.Spec.JsonData)
+				if err != nil {
+					return fmt.Errorf("error converting jsonData to a higher version: %v", err)
+				}
+				sink.Spec.ContentType = cloudevents.ApplicationJSON
+				sink.Spec.Data = string(msg)
+			}
+		}
+
+		// marshal and store v1beta1.PingSource.Spec into V1B1SpecAnnotationKey
+		// this is to help if we need to convert back to v1beta1.PingSource
+		v1beta1Spec, _ := json.Marshal(source.Spec)
+		annotations[V1B1SpecAnnotationKey] = string(v1beta1Spec)
+		sink.SetAnnotations(annotations)
+
+		return nil
+	default:
+		return apis.ConvertToViaProxy(ctx, source, &v1beta2.PingSource{}, sink)
+	}
 }
 
 // ConvertFrom implements apis.Convertible
-func (sink *PingSource) ConvertFrom(ctx context.Context, source apis.Convertible) error {
-	return fmt.Errorf("v1beta1 is the highest known version, got: %T", source)
+// Converts obj from a higher version into v1beta1.PingSource.
+func (sink *PingSource) ConvertFrom(ctx context.Context, obj apis.Convertible) error {
+	switch source := obj.(type) {
+	case *v1beta2.PingSource:
+		sink.ObjectMeta = source.ObjectMeta
+		sink.Status = PingSourceStatus{
+			SourceStatus: source.Status.SourceStatus,
+		}
+
+		// deep copy annotations to avoid mutation on source.ObjectMeta.Annotations
+		annotations := make(map[string]string)
+		for key, value := range source.GetAnnotations() {
+			annotations[key] = value
+		}
+
+		// v1beta2 objects originally converted from v1beta1 is expected to have this annotation
+		v1beta1Spec, ok := annotations[V1B1SpecAnnotationKey]
```

Comment:
IMO, the conversion should fail when the annotation is not present. Otherwise the spec looks incomplete.

Reply:
the storage version is v1beta1, if we fail it here then when we create v1beta2 PingSource the annotation should also be specified, for example the following will fail:

```bash
$ kubectl create -n pingsource-example -f - <<EOF
apiVersion: sources.knative.dev/v1beta2
kind: PingSource
metadata:
  name: test-ping-source-v1beta2-x3
spec:
  schedule: "*/1 * * * *"
  contentType: "text/plain"
  data: "beta2 beta2 beta2"
  sink:
    ref:
      apiVersion: v1
      kind: Service
      name: event-display
EOF
Error from server: error when creating "STDIN": conversion webhook for sources.knative.dev/v1beta2, Kind=PingSource failed: conversion failed to version v1beta1 for type [kind=PingSource group=sources.knative.dev version=v1beta2] -  V1B1SpecAnnotationKey does not exist
```

Reply:
Oops sorry I forgot the storage version is v1beta1. What feels strange is when getting the v1beta1 version of a v1beta2 instance, the spec is not what I would expect. 

My example is this one:
```yaml
apiVersion: sources.knative.dev/v1beta2
kind: PingSource
metadata:
  name: hello-no-wrapper
spec:
  schedule: "*/1 * * * *"
  data: "hello"
  sink:
    ref:
      apiVersion: v1
      kind: Service
      name: event-display
```

```
$ kubectl get pingsources.v1beta1.sources.knative.dev hello-no-wrapper -oyaml
...
metadata:
  annotations:
    pingsources.sources.knative.dev/v1beta2-spec: '{"sink":{"ref":{"kind":"Service","name":"event-display","apiVersion":"v1"}},"schedule":"*/1
      * * * *","data":"hello"}'
spec:
  schedule: '* * * * *'
  sink: {}
...
```

Not a biggie. I don't think getting a previous version is common.

Reply:
yes, this goes back to our original discussion on round-tripping, there's no deterministic way to know the expected v1beta1 spec by only looking at v1beta2 spec. 

so here v1beta1 just serves as an envelope for v1beta2 spec.